notebooks/02_model/hypertune_spark_deep_dive.ipynb

```diff
@@ -0,0 +1,983 @@
+      "text/plain": [
+       "<Figure size 1080x360 with 2 Axes>"
+      ]
+     },
+     "metadata": {
+      "needs_background": "light"
+     },
+     "output_type": "display_data"
+    }
+   ],
+   "source": [
+    "parameters = ['rank', 'reg']\n",
+    "cols = len(parameters)\n",
+    "f, axes = plt.subplots(nrows=1, ncols=cols, figsize=(15,5))\n",
+    "cmap = plt.cm.jet\n",
+    "for i, val in enumerate(parameters):\n",
+    "    xs = np.array([t['misc']['vals'][val] for t in trials.trials]).ravel()\n",
+    "    ys = [t['result']['loss'] for t in trials.trials]\n",
+    "    xs, ys = zip(*sorted(zip(xs, ys)))\n",
+    "    ys = np.array(ys)\n",
+    "    axes[i].scatter(xs, ys, s=20, linewidth=0.01, alpha=0.75, c=cmap(float(i)/len(parameters)))\n",
+    "    axes[i].set_title(val)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "It can be seen from the above plot that\n",
+    "* The actual impact of rank is in line with the intuition - the larger the value the better the result.\n",
+    "* It is interesting to see that the optimal value of reg is ~0.1. "
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Get the best model."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 18,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "als = ALS(\n",
+    "    rank=best[\"rank\"],\n",
+    "    regParam=best[\"reg\"],\n",
+    "    maxIter=15,\n",
+    "    implicitPrefs=False,\n",
+    "    alpha=0.1,\n",
+    "    coldStartStrategy='drop',\n",
+    "    nonnegative=False,\n",
+    "    seed=0,\n",
+    "    **HEADER_ALS\n",
+    ")\n",
+    "    \n",
+    "model_best_hyperopt = als.fit(train)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Tuning prameters against other metrics can be simply done by modifying the `objective` function. The following shows an objective function of how to tune \"precision@k\". Since `fmin` in `hyperopt` only supports minimization while the actual objective of the loss is to maximize \"precision@k\", `-precision` instead of `precision` is used in the returned value of the `objective` function."
```

Comment:
I think this is super powerful. This allows optimizing for rating or for ranking independently

Reply:
Agree and this is what we want. Support custom loss function is a very important feature to have.